server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -1032,12 +1076,148 @@ public int hashCode() {
         }
     }
 
-    private static class Int8HnswIndexOptions extends IndexOptions {
+    static class Int4HnswIndexOptions extends IndexOptions {
+        private final int m;
+        private final int efConstruction;
+        private final float confidenceInterval;
+
+        Int4HnswIndexOptions(int m, int efConstruction, Float confidenceInterval) {
+            super("int4_hnsw");
+            this.m = m;
+            this.efConstruction = efConstruction;
+            // The default confidence interval for int4 is dynamic quantiles, this provides the best relevancy and is
+            // effectively required for int4 to behave well across a wide range of data.
+            this.confidenceInterval = confidenceInterval == null ? 0f : confidenceInterval;
+        }
+
+        @Override
+        public KnnVectorsFormat getVectorsFormat() {
+            return new ES814HnswScalarQuantizedVectorsFormat(m, efConstruction, confidenceInterval, 4, true);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("type", type);
+            builder.field("m", m);
+            builder.field("ef_construction", efConstruction);
+            builder.field("confidence_interval", confidenceInterval);
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Int4HnswIndexOptions that = (Int4HnswIndexOptions) o;
+            return m == that.m && efConstruction == that.efConstruction && Objects.equals(confidenceInterval, that.confidenceInterval);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(m, efConstruction, confidenceInterval);
+        }
+
+        @Override
+        public String toString() {
+            return "{type="
+                + type
+                + ", m="
+                + m
+                + ", ef_construction="
+                + efConstruction
+                + ", confidence_interval="
+                + confidenceInterval
+                + "}";
+        }
+
+        @Override
+        boolean supportsElementType(ElementType elementType) {
+            return elementType != ElementType.BYTE;
+        }
+
+        @Override
+        boolean updatableTo(IndexOptions update) {
+            // TODO: add support for updating from flat, hnsw, and int8_hnsw and updating params
+            return false;
+        }
+
+        @Override
+        void validateDimension(int dim) {
+            if (dim % 2 != 0) {
+                throw new IllegalArgumentException("int4_hnsw only supports even dimensions; provided=" + dim);
+            }
+        }
+    }
+
+    static class Int4FlatIndexOptions extends IndexOptions {
```

Comment:
Nit: Would it make sense for this class to extend `Int8FlatIndexOption`, or extract a common base class?